lib/pbench/server/api/resources/generate_key.py

```diff
@@ -0,0 +1,147 @@
+from flask.wrappers import Request, Response
+from flask_restful import abort
+import jwt
+from sqlalchemy.exc import IntegrityError, SQLAlchemyError
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import (
+    APIAbort,
+    ApiAuthorizationType,
+    ApiBase,
+    ApiContext,
+    ApiMethod,
+    ApiParams,
+    ApiSchema,
+)
+import pbench.server.auth.auth as Auth
+from pbench.server.database.models.api_key import APIKey
+from pbench.server.database.models.audit import OperationCode
+
+
+def generate_key(time_delta: timedelta, user):
+    user_obj = user.get_json()
+    current_utc = datetime.now(timezone.utc)
+    expiration = current_utc + time_delta
+    payload = {
+        "iat": current_utc,
+        "exp": expiration,
+        "user_id": user_obj["id"],
+        "username": user_obj["username"],
+        "audience": Auth.oidc_client.client_id,
+    }
+    try:
+        generated_api_key = jwt.encode(
+            payload, current_app.secret_key, algorithm=Auth._TOKEN_ALG_INT
+        )
+    except (
+        jwt.InvalidIssuer,
+        jwt.InvalidIssuedAtError,
+        jwt.InvalidAlgorithmError,
+        jwt.PyJWTError,
+    ):
+        current_app.logger.exception(
+            "Could not encode the JWT api_key for user: {}", user
+        )
+        abort(
+            HTTPStatus.INTERNAL_SERVER_ERROR,
+            message="INTERNAL ERROR",
+        )
+    return generated_api_key, current_utc, expiration
+
+
+class GenerateKey(ApiBase):
+    def __init__(self, config: PbenchServerConfig):
+        super().__init__(
+            config,
+            ApiSchema(
+                ApiMethod.POST,
+                OperationCode.CREATE,
+                authorization=ApiAuthorizationType.NONE,
+            ),
+        )
+        self.server_config = config
+
+    def _post(
+        self, params: ApiParams, request: Request, context: ApiContext
+    ) -> Response:
+        """
+        Post request for generating a new persistent API key.
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+
+        :return:
+            Success: 201 with empty payload
+            Failure: <status_Code>,
+                    response_object = {
+                        "message": "failure message"
+                    }
+        """
+        user = Auth.token_auth.current_user()
+        api_key_expire_duration = self.server_config.get(
+            "openid", "api_key_expiration_duration"
+        )
+
+        if user:
+            try:
+                new_key, current_utc, expiration = generate_key(
+                    time_delta=timedelta(minutes=int(api_key_expire_duration)),
+                    user=user,
+                )
+                try:
+                    key = APIKey(
+                        api_key=new_key, created=current_utc, expiration=expiration
+                    )
+                    user.update(api_key=key)
```

Comment:
Yeah, I think we want to support multiple keys for a user, which might be used (and selectively revoked) for separate purposes. E.g., our Jenkins connection to GitHub is authenticated by Peter's API key, and when looking at a list of your tokens it might be nice to see "this one was for the <foo> CI, this one was for cluster performance data, etc."

So maybe replace the expiration timestamp with a *string* value which could be specified as a JSON body (e.g., `{"name": "CI key"}`), which would be retrieved via `GET` along with the key, and displayed in the UI to help with management?

Reply:
Good idea. We can implement this one. I need some clarifications on this, whether these key value JSON attributes be stored only in the DB while saving `api_key`?

>I don't know if we want to include this enhancement in this PR, but I suppose it's pretty simple

I'll add these in the next commit if there are changes requested or else I will raise a new PR for this

Reply:
I believe that Dave's suggestion is that the `POST` method accept a JSON object as a payload from which the Pbench Server will extract a string (e.g., by looking for a key, `"name"`), and it will store that string in the database row with the API key value and the `User` reference.  We can worry later about how this string might be retrieved or updated.

However, if what we're doing is enhancing an existing method to now accept a payload (which we cannot yet `GET` or `PUT`), perhaps that would be better to do in a subsequent PR.